src/crypto/equihash.cpp

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2016 Jack Grigg
+// Copyright (c) 2016 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "crypto/equihash.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+
+void validate_params(int n, int k)
+{
+    if (k>=n) {
+        std::cerr << "n must be larger than k\n";
+        throw invalid_params();
+    }
+    if (n % 8 != 0) {
+        std::cerr << "Parameters must satisfy n = 0 mod 8\n";
+        throw invalid_params();
+    }
+    if ((n/(k+1)) % 8 != 0) {
+        std::cerr << "Parameters must satisfy n/(k+1) = 0 mod 8\n";
+        throw invalid_params();
+    }
+}
+
+StepRow::StepRow(unsigned int n, const CSHA256& hasher, uint32_t i) :
+        hash {new unsigned char[n/8]},
+        len {n/8},
+        indices {i}
+{
+    unsigned char tmp[hasher.OUTPUT_SIZE];
+    CSHA256(hasher)
+            .Write((unsigned char*) &i, sizeof(uint32_t))
+            .Finalize(tmp);
+    // TODO generate correct size instead?
+    std::copy(tmp, tmp+(n/8), hash);
+
+    assert(indices.size() == 1);
+}
+
+StepRow::~StepRow()
+{
+    delete[] hash;
+}
+
+StepRow::StepRow(const StepRow& a) :
+        hash {new unsigned char[a.len]},
+        len {a.len},
+        indices(a.indices)
+{
+    for (int i = 0; i < len; i++)
+        hash[i] = a.hash[i];
+}
+
+StepRow& StepRow::operator=(const StepRow& a)
+{
+    unsigned char* p = new unsigned char[a.len];
+    for (int i = 0; i < a.len; i++)
+        p[i] = a.hash[i];
+    delete[] hash;
+    hash = p;
+    len = a.len;
+    indices = a.indices;
+    return *this;
+}
+
+StepRow& StepRow::operator^=(const StepRow& a)
+{
+    if (a.len != len)
+        throw 1; // TODO throw better error
+    unsigned char* p = new unsigned char[len];
+    for (int i = 0; i < len; i++)
+        p[i] = hash[i] ^ a.hash[i];
+    delete[] hash;
+    hash = p;
+    indices.reserve(indices.size() + a.indices.size());
+    indices.insert(indices.end(), a.indices.begin(), a.indices.end());
+    std::sort(indices.begin(), indices.end());
+    return *this;
+}
+
+bool StepRow::IsZero()
+{
+    char res = 0;
+    for (int i = 0; i < len; i++)
+        res |= hash[i];
+    return res == 0;
+}
+
+bool HasCollision(StepRow& a, StepRow& b, int i, int l)
+{
+    bool res = true;
+    for (int j = (i - 1) * l / 8; j < i * l / 8; j++)
+        res &= a.hash[j] == b.hash[j];
+    return res;
+}
+
+bool DistinctIndices(const StepRow& a, const StepRow& b)
+{
+    unsigned int i = 0, j = 0;
+    while (i < a.indices.size() && j < b.indices.size()) {
+        while (i < a.indices.size() && a.indices[i] < b.indices[j])
+            i++;
+        while (j < b.indices.size() && a.indices[i] > b.indices[j])
+            j++;
+        if (a.indices[i] == b.indices[j])
+            return false;
```

Comment:
This can be made simpler as well as fixing the bug above:
```
unsigned int i = 0;
for (unsigned int j = 0; j < b.indices.size(); j++) {
    while (a.indices[i] < b.indices[j]) {
        i++;
        if (i == a.indices.size()) { return true; }
    }
    assert a.indices[i] >= b.indices[j];
    if (a.indices[i] == b.indices[j]) { return false; }
}
return true;
```